packages/gas-fee-controller/CHANGELOG.md

```diff
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [1.0.1]
+### Changed
+- Relax dependencies on `@metamask/base-controller`, `@metamask/controller-utils`, and `@metamask/network-controller` (use `^` instead of `~`) ([#998](https://github.com/MetaMask/controllers/pull/998))
+- Align with changes to `@metamask/network-controller`'s `NetworkState` type (`provider` -> `providerConfig`) ([#995](https://github.com/MetaMask/controllers/pull/995))
```

Comment:
Hmm. This one doesn't change anything for the constructor parameters, but it changes the expectations for the controller messenger. It now expects `NetworkController:providerConfigChange` rather than `NetworkController:providerChange`.

Seems like a breaking change, as it effectively requires the latest version of the network controller to be used unless the `onNetworkStateChange` callback is passed in.

Reply:
> Seems like a breaking change, as it effectively requires the latest version of the network controller to be used unless the `onNetworkStateChange` callback is passed in.

It looks like there is existing logic which checks to see if `onNetworkStateChange` is passed and if so uses that instead of calling `NetworkController:getProviderConfigChange`. So in the context of this package, it doesn't seem like it would be a breaking change?

However, I suppose if you were updating this package and were calling `NetworkController:getProviderConfig` yourself outside of this package, then your code would break, or wouldn't do what you expect, anyway. So what if this is the following?

``` suggestion
- **BREAKING:** Bump `@metamask/network-controller` to 2.0.0. See the changelog for this package for more information on breaking changes.
```

Reply:
Hmm. The problem is that the network controller being messaged isn't necessarily the same as the one brought in as a dependency. The project could have a different version, and have that different version configured as part of the controller messenger.

This is breaking not just in that the network dependency changed. It's breaking in that it now expects v2 of the network controller to be the one in the controller messenger.

Reply:
> The problem is that the network controller being messaged isn't necessarily the same as the one brought in as a dependency. 

Okay, so what you're saying is that the project could have:

* `@metamask/base-controller` 1.1.1
* `@metamask/gas-fee-controller` 1.0.1, which depends on:
  * `@metamask/base-controller` 1.1.1
  * `@metamask/network-controller` 2.0.0
* `@metamask/network-controller` 1.0.0, which depends on:
  * `@metamask/base-controller` 1.1.1

And the project could have set things up this way:

``` javascript
import { ControllerMessenger } from "@metamask/base-controller";
import { GasFeeController } from "@metamask/gas-fee-controller";
import { NetworkController } from "@metamask/network-controller";

const messenger = new ControllerMessenger(...);
const gasFeeController = new GasFeeController({ messenger })
const networkController = new NetworkController({ messenger })
```

So now GasFeeController attempts to call `NetworkController:getProviderConfigChange` via the messenger, but NetworkController doesn't know how to respond to that, so nothing happens.

I have two thoughts:

1. When using `^` to specify versions, would Yarn dedupe the dependencies? For instance if we instead we had:

    * `@metamask/base-controller` ^1.1.1
    * `@metamask/gas-fee-controller` ^1.0.1, which depends on:
      * `@metamask/base-controller` ^1.1.1
      * `@metamask/network-controller` ^2.0.0
    * `@metamask/network-controller` ^1.0.0, which depends on:
      * `@metamask/base-controller` ^1.1.1

    would Yarn see that ^2.0.0 is satisfied by ^1.0.0, install 2.x, and use it for both `gas-fee-controller` and `network-controller`?

2. It seems like this problem — that you can have a situation where you have a dependency installed directly at one version and indirectly at another — is inherent in NPM's design and will always be present when it comes to using multiple controllers together no matter what we do. Releasing a new major version won't make that problem go away, nor is that problem only present when we introduce what we could call breaking changes. The same breakage that we describe here could happen even if we release a minor version. For instance, if we released a new version of `gas-fee-controller` that bumped the version of `network-controller` because we added a new event, but that event wasn't present in a previous version of `network-controller` which the project was directly relying upon, then attempting to use that event would fail, too. Wouldn't it?

Reply:
Updated d73845f...4fdb9c2.

Reply:
>would Yarn see that ^2.0.0 is satisfied by ^1.0.0, install 2.x, and use it for both gas-fee-controller and network-controller?

Nope; those are incompatible ranges. Dedupe only handles in-range duplication.

Though we could solve this in both the extension and mobile using yarn constraints, when they are migrated to yarn v3.

Reply:
This is indeed surprising and weird. It's kinda our fault rather than npm's, since we're introducing this dependency via the controller messenger that npm isn't aware of.

Perhaps adding it as a peer dependency would help solve this? Otherwise I agree with your server analogy, we need some sort of handshake process to declare expectations upfront.